### PR TITLE
Clean up the stack implementations

### PR DESCRIPTION
**Pull request by Amanieu**
Changes:
- ~~The valgrind stuff is moved out of the generators and into the stack implementations.~~
- `Stack` is now an unsafe trait since it has a contract that must be upheld by an implementation.
- `SliceStack` will now automatically align a slice on construction. However this means that its fields are now private.
- Various cleanups and moving code around.

### CODE REVIEW

**Comment by whitequark** [comment]
First, can you please separate the "clean ups" into distinct commits?

> The valgrind stuff is moved out of the generators and into the stack implementations.

Uh, how about no. Now anyone implementing a stack externally will now have to 1) implement the Valgrind interface 2) implement interfaces of any other debugging tools we may be interested in the future 3) predicate on the architecture-specific availability of Valgrind themselves. This significantly decreases ergonomics (and also it's not a cleanup, it creates duplication).

> SliceStack will now automatically align a slice on construction. However this means that its fields are now private.

Again, how about no. This makes it completely useless for its original purpose, that is using a `static mut` for a stack, since you cannot construct it. You could as well just remove SliceStack entirely now.

**Comment by Amanieu** [comment]
> Again, how about no. This makes it completely useless for its original purpose, that is using a static mut for a stack, since you cannot construct it. You could as well just remove SliceStack entirely now.

Can't you just use a `static mut [u8; N]` and create a `SliceStack` from that?

> Uh, how about no. Now anyone implementing a stack externally will now have to 1) implement the Valgrind interface 2) implement interfaces of any other debugging tools we may be interested in the future 3) predicate on the architecture-specific availability of Valgrind themselves. This significantly decreases ergonomics (and also it's not a cleanup, it creates duplication).

The motivation here was to associate the debug info with the stacks rather than with the generators. Consider the case where a stack is reused for multiple generators, this would cause the stack to be registered and de-registered multiple times.

File: src/stack/slice_stack.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    data: &'a mut [u8],
+    id: StackId,
+}
+
+impl<'a> SliceStack<'a> {
+    /// Creates a `SliceStack` from an existing slice.
+    ///
+    /// This function will automatically align the slice to make it suitable for
+    /// use as a stack. However this function may panic if the slice is smaller
+    /// than `STACK_ALIGNMENT`.
```
Comment (whitequark) [inline src/stack/slice_stack.rs]: Where does it panic in release builds?

**Comment by whitequark** [comment]
> The motivation here was to associate the debug info with the stacks rather than with the generators. Consider the case where a stack is reused for multiple generators, this would cause the stack to be registered and de-registered multiple times.

I don't see how this is a problem. You're sacrificing ergonomics in a misguided attempt to improve efficiency of a feature only useful for debugging in the first place. (And we should also gate the valgrind stuff behind `#[cfg(build = "debug")]`.)

> Can't you just use a static mut [u8; N] and create a SliceStack from that?

Hm, I tried that and encountered some sort of problem, but I don't recall it now. Disregard that comment then.

That said, if SliceStack::new didn't mess with valgrind then it could be a const fn.

File: src/stack/slice_stack.rs (hunk 1 above)
Comment (Amanieu) [inline src/stack/slice_stack.rs]: This may panic even in release builds: `&mut slice[offset..(offset + adjusted_len)]`

Comment (whitequark) [inline src/stack/slice_stack.rs]: Worth a comment, I think.

**Comment by Amanieu** [comment]
I reverted the valgrind changes.

**Comment by whitequark** [comment]
LGTM. SliceStack's `new` can be made a `const fn` later if necessary, or the field exposed anyway, or something to that degree. As it is it is definitely nicer than what I had, so your change there is a net win.

**Comment by whitequark** [comment]
There's no real point in making `Stack` unsafe but since people persistently point that out I suppose there is no real harm either.

**Comment by Amanieu** [comment]
I don't think `SliceStack::new` can be made a const fn since it needs to inspect the address of the slice to ensure that it is properly aligned.

Also, publicly exposing the field would allow safety violations (in particular, the alignment restriction won't be enforced).

**Comment by whitequark** [comment]
> I don't think SliceStack::new can be made a const fn since it needs to inspect the address of the slice to ensure that it is properly aligned.

Right.

> Also, publicly exposing the field would allow safety violations (in particular, the alignment restriction won't be enforced).

Using SliceStack is already unsafe since it provides no guard page; exposing the field could be seen as inelegant design since it lets you violate what's supposed to be an invariant, but it does not add any more unsafety.

File: src/stack/slice_stack.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        let ptr = slice.as_ptr() as usize;
+        let adjusted_ptr = (ptr + ::STACK_ALIGNMENT - 1) & !(::STACK_ALIGNMENT - 1);
+        let offset = adjusted_ptr - ptr;
+        if offset > slice.len() {
+            panic!("SliceStack too small");
+        }
+        let adjusted_len = (slice.len() - offset) & !(::STACK_ALIGNMENT - 1);
+
+        // This may panic if offset is larger than the size of the slice
+        SliceStack(&mut slice[offset..(offset + adjusted_len)])
```
Comment (whitequark) [inline src/stack/slice_stack.rs]: Please group these two panics into one explicit one, and write a test for both conditions.

Comment (Amanieu) [inline src/stack/slice_stack.rs]: The only panic is the explicit one. `adjusted_len` is guaranteed to be less than `slice.len() - offset`.

Comment (whitequark) [inline src/stack/slice_stack.rs]: So the comment is wrong now?

Comment (Amanieu) [inline src/stack/slice_stack.rs]: Oops, forgot about that one. Fixed.

Comment (whitequark) [inline src/stack/slice_stack.rs]: And the test?

Comment (Amanieu) [inline src/stack/slice_stack.rs]: Test added.

File: src/stack/slice_stack.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        unsafe { self.limit().offset(len) }
+    }
+
+    #[inline(always)]
+    fn limit(&self) -> *mut u8 {
+        self.0.as_ptr() as *mut u8
+    }
+}
+
+#[cfg(test)]
```
Comment (whitequark) [inline src/stack/slice_stack.rs]: These tests don't poke into any implementation details as far as I see, so shouldn't they go to `tests/`?

Comment (Amanieu) [inline src/stack/slice_stack.rs]: Well I feel that the precise behavior of whether it panics or not is a bit of an implementation detail. The only thing the interface guarantees is that it won't panic if the slice is larger than `STACK_ALIGNMENT`, and it *may* panic if it is smaller.

Comment (whitequark) [inline src/stack/slice_stack.rs]: That's a weird contract...

**Comment by Amanieu** [comment]
I moved the tests into `tests/stack.rs`.

**Pull request closed without merging**